### DISCUSSION TITLE
2.x JSON Parser - Error handling and Defaults

### DISCUSSION THREAD

**Discussion by danbo** [savonet/liquidsoap#2534] [User Support]
https://www.liquidsoap.info/doc-dev/json.html

I'm looking at supplying default values when a parse fails I'm not sure if that's currently possible with the new parser, it doesn't understand if I give defaults in the fields or type annotations:

```
let json.parse ( x : {
    foo: float **= 0.0**,
    "gni gno" as gni_gno: bool,
    nested: {
      tuple: (_ * float),
      list: [float],
      nullable_list: [int?],
      object_as_list: [(string * float)] as json.object,
      "arbitrary object key ✨" as arbitrary_object_key: bool,
      not_present: bool?
    }
  }) = data
```

What is the best way to handle this? It's not clear in the documentation. The only thing I can think of is using a try/catch, but then I have additional code to assign fallback values through ref assignments.

**Comment by toots** [comment]
Excellent question.

For now, I don't think that you can pass inline default values. You can, however, parse something as optional and then give it a default value:

```ruby
let json.parse ({foo, bla} : {foo: int, bla: string?}) = '{"foo":123}'
- : unit = ()

let bla = bla ?? "default"
- : unit = ()

bla
- : string = "default"
```

**Reply by danbo** [reply]
Can you show an example of the best way to set default if there was a fetch failure, ie
```
let json.parse ({foo, bla} : {foo: int, bla: string?}) = '**{}**'
```
